### PR TITLE
issue/72 - Implement a reminder system for the calendar tasks closes #72

### LINKED ISSUE #72: Implementing a reminder system for the calendar tasks
### User Story: 

As a software developer, I want to set up reminders for tasks associated with my calendar app so that the app users can get appropriate notifications.

### Acceptance Criteria:

1) A seamless implementation of a reminder system on the calendar page.

### Why is the feature required?
So that the users of the app can get appropriate notifications for their scheduled tasks.

### PR DESCRIPTION
**Pull request by Minus20Five**
* Set the new task reminderDate in task scheduler.
* Implemented `reminder-service` which runs a repeat task to poll for new reminders and fire them off via web push. Currently, we are polling every minute which means notifications will be delayed by 1 minute at most.
* On the client side we first request permissions for notifications.
* If permission granted, register a service worker to handle notification events and show the notification popup.
* Also on the client side, we pass a subscription object to the server after auth redirect so that the server can map each client to its subscription.
* Updated .env for client and server containing VAPID keys. 
* Updated unit tests.

image

- [x] The pull request is complete according to the following criteria:
  - [x] Acceptance criteria have been met
  - [x] The documentation is kept up-to-date
  - [x] Comprehensive tests (if applicable) have been generated and all pass.
  - [x] The pull request describes the changes that have been made, and enough information is present in the description for any developer to understand what has changed
  - [x] Commits have been squashed (or will be on merge).
  - [x] The branch name is descriptive and follows the pull request title format : {issue/bug...}/(Issue Number) - Name of issue. E.g bug/30-Fix-Project
  - [x] The pull request title is of the following format : {issue/bug...}/(Issue Number) - Name of issue. E.g bug/30-Fix-Project
  - [x] The description uses [github syntax](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) to link to the issue. E,g Resolves se701g2/Doto#{Number}
  - [x] At least two reviewers assigned. One of which must be the assigner of the issue.
  - [x] If there are merge conflicts, run git rebase as opposed to git merge with master.

### CODE REVIEW

File: doto-frontend/public/reminderWorker.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+self.addEventListener("push", ev => {
+    const data = ev.data.json();
+    self.registration.showNotification(data.title, {
+        body: data.description,
```
Comment (HarrisonLeach1) [inline doto-frontend/public/reminderWorker.js:4]: Currently, the notification just displays the task title and description, but the user might not know what to make of it.

We could modify the code here to display something like: 
REMINDER: \<task name\> due/happening at \<time\>

Might be a better UX.

**Comment by Matteas-Eden** [comment]
Yeah so I got an error when testing. I could access the calendar page, but after about a second it would come up with the below error. Refreshing the page resulted in the same behaviour - calendar page for a second, then error.

**OS**: Arch Linux
**Browser**: Firefox v75
**Version**: b60c4be

PR-241_Error

File: doto-frontend/public/reminderWorker.js (hunk 1 above)
Comment (Minus20Five) [reply doto-frontend/public/reminderWorker.js]: Yeah thats a good idea. I've added a commit to format the start date in the notification title.

**Comment by Minus20Five** [comment]
> Yeah so I got an error when testing. I could access the calendar page, but after about a second it would come up with the below error. Refreshing the page resulted in the same behaviour - calendar page for a second, then error.
> 
> **OS**: Arch Linux
> **Browser**: Firefox v75
> **Version**: [b60c4be](https://github.com/se701g2/Doto/commit/b60c4be634c54059d0ed2bae51d17b78abf6b12d)
> 
> PR-241_Error

This seems like a [Firefox specific issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1429714). I don't have Firefox installed so I've only tested on Chrome. Do you mind following the instructions in the link and check your cookie preferences are set to "keep until they expire" and not "keep until I close firefox" and see if that changes anything.

**Comment by Minus20Five** [comment]
@all-contributors please add @Minus20Five for code, review, design, doc, test

**Pull request merged**